src/main/java/org/opensearch/knn/training/VectorReader.java

```diff
@@ -0,0 +1,162 @@
+     */
+    public void read(String indexName, String fieldName, int maxVectorCount, int searchSize,
+                     Consumer<List<Float[]>> vectorConsumer, ActionListener<SearchResponse> listener) {
+
+        ValidationException validationException = null;
+
+        // Validate arguments
+        if (maxVectorCount <= 0) {
+            validationException = new ValidationException();
+            validationException.addValidationError("maxVectorCount must be >= 0");
+        }
+
+        if (searchSize > 10000 || searchSize <= 0) {
+            validationException = validationException == null ? new ValidationException() : validationException;
+            validationException.addValidationError("searchSize must be > 0 and <= 10000");
+        }
+
+        IndexMetadata indexMetadata = indicesService.clusterService().state().metadata().index(indexName);
+        if (indexMetadata == null) {
+            validationException = validationException == null ? new ValidationException() : validationException;
+            validationException.addValidationError("index \"" + indexName + "\" does not exist");
+            throw validationException;
+        }
+
+        MappedFieldType fieldType = indicesService.indexServiceSafe(indexMetadata.getIndex()).mapperService()
+                .fieldType(fieldName);
+
+        if (!(fieldType instanceof KNNVectorFieldMapper.KNNVectorFieldType)) {
+            validationException = validationException == null ? new ValidationException() : validationException;
+            validationException.addValidationError("field \"" + fieldName + "\" must be of type KNNVectorFieldType");
+        }
+
+        if (validationException != null) {
+            throw validationException;
+        }
+
+        // Start reading vectors from index
+        ActionListener<SearchResponse> vectorReaderListener = getVectorReaderListener(fieldName,
+                maxVectorCount, 0, vectorConsumer, listener);
+
+        createSearchRequestBuilder(indexName, fieldName, Integer.min(maxVectorCount, searchSize))
+                .execute(vectorReaderListener);
+    }
+
+    @SuppressWarnings("unchecked")
+    private ActionListener<SearchResponse> getVectorReaderListener(String fieldName,
+                                                                   final int maxVectorCount,
+                                                                   final int collectedVectorCount,
+                                                                   Consumer<List<Float[]>> vectorConsumer,
+                                                                   ActionListener<SearchResponse> listener) {
+        return ActionListener.wrap(searchResponse -> {
+            // Get the vectors from the search response
+            // Either add the entire set of returned hits, or maxVectorCount - collectedVectorCount hits
+            SearchHit[] hits = searchResponse.getHits().getHits();
```

Comment:
Is there a time or number of scrolls limit? or we can run this endlessly?

Reply:
A limit is imposed by `maxVectorCount`. This can be controlled by user to make sure we do not make too many queries.